### PR TITLE
Remove some unsafe

### PR DESCRIPTION
**Pull request by cpcloud**
This PR removes a number of unsafe that calls that appear to be unnecessary.

### CODE REVIEW

File: src/reader.rs [hunk 1]
```diff
@@ -99,12 +99,9 @@ impl<R: Read> Block<R> {
     fn fill_buf(&mut self, n: usize) -> Result<(), Error> {
         // We don't have enough space in the buffer, need to grow it.
         if n >= self.buf.capacity() {
-            self.buf.reserve(n);
+            self.buf.resize(n, 0);
```
Comment (poros) [inline src/reader.rs]: I think this one may require a `Clone` operation. Can you please run the benchmark both for master and this branch and post the results? Just to double check that this doesn't cause any evident performance hit.

Comment (cpcloud) [reply src/reader.rs]: It looks like there are some sizable wins when reading large-schema'd files, with some small (?) losses in reading small schema'd files:

```
 name                                     master.bench ns/iter  remove-some-unsafe.bench ns/iter  diff ns/iter   diff %  speedup
 bench_big_schema_json_read_10000_record  25,529,478            22,951,785                          -2,577,693  -10.10%   x 1.11
 bench_big_schema_read_100000_record      101,731,787           77,780,801                         -23,950,986  -23.54%   x 1.31
 bench_big_schema_read_10000_record       10,741,198            7,820,490                           -2,920,708  -27.19%   x 1.37
 bench_big_schema_read_100_record         147,585               126,039                                -21,546  -14.60%   x 1.17
 bench_big_schema_read_1_record           42,689                42,443                                    -246   -0.58%   x 1.01
 bench_big_schema_write_10000_record      2,282,573             2,302,998                               20,425    0.89%   x 0.99
 bench_big_schema_write_100_record        25,475                25,045                                    -430   -1.69%   x 1.02
 bench_big_schema_write_1_record          4,583                 4,476                                     -107   -2.33%   x 1.02
 bench_big_schema_write_record            610                   652                                         42    6.89%   x 0.94
 bench_file_quickstop_null                2,972,934             2,234,466                             -738,468  -24.84%   x 1.33
 bench_small_schema_read_10000_record     1,033,114             1,163,512                              130,398   12.62%   x 0.89
 bench_small_schema_read_100_record       17,882                19,065                                   1,183    6.62%   x 0.94
 bench_small_schema_read_1_record         7,645                 7,472                                     -173   -2.26%   x 1.02
 bench_small_schema_write_10000_record    297,972               284,443                                -13,529   -4.54%   x 1.05
 bench_small_schema_write_100_record      4,934                 4,641                                     -293   -5.94%   x 1.06
 bench_small_schema_write_1_record        1,694                 1,602                                      -92   -5.43%   x 1.06
 bench_small_schema_write_record          99                    98                                          -1   -1.01%   x 1.01
```

**Comment by cpcloud** [comment]
It looks like there are some sizable wins when reading large-schema'd files, with some small (?) losses in reading small schema'd files:

```
 name                                     master.bench ns/iter  remove-some-unsafe.bench ns/iter  diff ns/iter   diff %  speedup
 bench_big_schema_json_read_10000_record  25,529,478            22,951,785                          -2,577,693  -10.10%   x 1.11
 bench_big_schema_read_100000_record      101,731,787           77,780,801                         -23,950,986  -23.54%   x 1.31
 bench_big_schema_read_10000_record       10,741,198            7,820,490                           -2,920,708  -27.19%   x 1.37
 bench_big_schema_read_100_record         147,585               126,039                                -21,546  -14.60%   x 1.17
 bench_big_schema_read_1_record           42,689                42,443                                    -246   -0.58%   x 1.01
 bench_big_schema_write_10000_record      2,282,573             2,302,998                               20,425    0.89%   x 0.99
 bench_big_schema_write_100_record        25,475                25,045                                    -430   -1.69%   x 1.02
 bench_big_schema_write_1_record          4,583                 4,476                                     -107   -2.33%   x 1.02
 bench_big_schema_write_record            610                   652                                         42    6.89%   x 0.94
 bench_file_quickstop_null                2,972,934             2,234,466                             -738,468  -24.84%   x 1.33
 bench_small_schema_read_10000_record     1,033,114             1,163,512                              130,398   12.62%   x 0.89
 bench_small_schema_read_100_record       17,882                19,065                                   1,183    6.62%   x 0.94
 bench_small_schema_read_1_record         7,645                 7,472                                     -173   -2.26%   x 1.02
 bench_small_schema_write_10000_record    297,972               284,443                                -13,529   -4.54%   x 1.05
 bench_small_schema_write_100_record      4,934                 4,641                                     -293   -5.94%   x 1.06
 bench_small_schema_write_1_record        1,694                 1,602                                      -92   -5.43%   x 1.06
 bench_small_schema_write_record          99                    98                                          -1   -1.01%   x 1.01
```

**Comment by cpcloud** [comment]
The only thing that looks meaningfully regressive that doesn't look like noise is:

```
 bench_small_schema_read_10000_record 1,033,114 1,163,512 130,398 12.62% x 0.89
```

File: src/reader.rs (hunk 1 above)
Comment (cpcloud) [reply src/reader.rs]: Only `T` must satisfy `Clone` here, and here `T` reduces to a `u8` whose `clone` is extremely cheap.

**Comment by poros** [comment]
Mmm, I would say that it's still alright and I'd really like to get rid of all the unsafe code we can. Let's go ahead :)
Thanks again.

**Pull request merged**